### ISSUE TITLE
mask() and rasterize() creates artificial holes

### ISSUE THREAD

**Issue by yPennylane** [rspatial/raster#60]
I have a spatialPolygonsDataFrame consisting of 3 polygons. The third polygon has the same shape as the first but has a hole where the second polygon is located.

I created a raster and a spatialPolygonsDataFrame to demonstrate the issue:

    # create rasters and store them in a list
    r1 <- raster(xmn=1, xmx=5, ymn=1, ymx=5, nrows=4, ncols=4)
    r1[] <- 1:length(r1)
    
    # create SpatialPolygonsDataFrame
    Sr1 = Polygon(cbind(c(1,5,4,1,1),c(1,2,5,4,1)))
    Sr2 = Polygon(cbind(c(2,4,3,2),c(3,2,4,3)))
    SpP = SpatialPolygons(list(Polygons(list(Sr1), "s1"), Polygons(list(Sr2), "s2")), 
                          1:2)
    dat = data.frame(ID = c("s1", "s2"), value = c("a", "b"))
    row.names(dat) <- c("s1", "s2")
    p <- SpatialPolygonsDataFrame(SpP, data = dat, 
                                  match.ID = TRUE)
    
    AddHoleToPolygon <-function(poly,hole){
      # invert the coordinates for Polygons to flag it as a hole
      coordsHole <-  hole@polygons[[1]]@Polygons[[1]]@coords
      newHole <- Polygon(coordsHole,hole=TRUE)
    
      # punch the hole in the main poly
      listPol <- poly@polygons[[1]]@Polygons
      listPol[[length(listPol)+1]] <- newHole
      punch <- Polygons(listPol,poly@polygons[[1]]@ID)
    
      # make the polygon a SpatialPolygonsDataFrame as the entry
      new <- SpatialPolygons(list(punch),proj4string=poly@proj4string)
      new <- SpatialPolygonsDataFrame(new,data=as(poly,"data.frame"))
    
      return(new)
    }
    
    punchedPoly <-AddHoleToPolygon(p[1,],p[2,])
    
    p1 <- rbind(p, punchedPoly, makeUniqueIDs = TRUE)
    p1 <- p1[2:3,]

When I use `mask()` to "crop" the raster r1, then the hole is created, although the triangular polygon has a value and indeed is not a real hole. But it gets "overridden" by the third polygon with the hole:

    masked_hole <- mask(r1, p1)
    plot(masked_hole)

When I change the order of the polygons, then no hole is created:

m3 <- mask(r1, p1[c(2,1),])
plot(m3)

The function rasterize is affected in the same manner:

    r2 <- rasterize(p1, r1, field = "value")
    plot(r2)
    r3 <- rasterize(p1[c(2,1),], r1, field = "value")
    plot(r3)

In my real data I have holes where there are no "filling" polygons and those ones I want to keep as holes. 

I did a workaround of this issue by ordering the polygons by area (`shape@polygons[[i]]@area`, descending) and then used the new object as input for rasterize (and input for mask, respectively).

Could the behaviour of `mask()` and `rasterize` be fixed by assuring that "fillings" aren't overwritten by polygons with holes?

(stackoverflow: https://stackoverflow.com/questions/56759025/r-mask-and-rasterize-with-spatialpolygonsdataframe-having-holes)

**Comment by rhijmans** [comment]
Thanks, and yes, that needs to be fixed.  Other possible work arounds: 

#1
rf <- rasterize(p1, r1, field =1, fun=max)

#2 
x = union(p1)
ru <- rasterize(x, r1)

**Comment by rhijmans** [comment]
I believe this is now fixed (raster version 2.9-19)

```
library(raster)
Sr1 = cbind(c(1,5,4,1,1),c(1,2,5,4,1))
Sr2 = cbind(c(2,4,3,2),c(3,2,4,3))
# polygon with hole first
p1 <- spPolygons(list(Sr1, Sr2), Sr2, attr=data.frame(v=c("a", "b")))
# polygon with hole last
p2 <- p1[2:1,]
# only the polygon with hole
p0 <- p1[1,] 
plot(p1, col=c("red", "blue"))

r <- raster(p1, res=.1)
values(r) <- 1:ncell(r)

m0 <- mask(r, p0)
m1 <- mask(r, p1)
m2 <- mask(r, p2)
mm <- stack(m0, m1, m2)
names(mm) <- c("hole-only", "hole-first", "hole-last")
spplot(mm)

r0 <- rasterize(p0, r)
r1 <- rasterize(p1, r)
r2 <- rasterize(p2, r)
rr <- stack(r0, r1, r2)
names(rr) <- c("hole-only", "hole-first", "hole-last")
spplot(rr)

```
